### ISSUE TITLE
Password not always cleared when locking the database

### ISSUE THREAD

**Issue by nhunsperger** [PhilippC/keepass2android#829]
The master password is not cleared if the screen is rotated during the unlock operation.  This results in a situation where the DB can be unlocked later without knowing the master password.  Worse, this can expose the master password by clicking the "eye".

Steps to reproduce:
1) Enter Password
2) Tap "Unlock"
3) While "Working..." dialog is showing, rotate screen (OK to rotate back)
4) Tap "Lock" icon
5) Notice password still present
6) Unlock DB without entering the password (or click the eye to learn the master password)

Note: This is easiest to exploit if using KDBX4 format tuned for desktop unlocking.  Locking the DB via device screen lock is also affected.  Password stays entered when app is sent to background, other apps used, and later brought to foreground.

Version: 1.06g
Variant: Offline
Device: Samsung Galaxy S7
Android: 8.0.0

**Comment by dragoangel** [comment]
Did you enable clean entered master password in application settings?

**Comment by nhunsperger** [comment]
> Did you enable clean entered master password in application settings?

No, but that setting should be unrelated.  "Clear the entered master password when leaving the screen *without unlocking* the database".  I unlocked the database.

Under normal use (no screen rotation), with or without this setting enabled, the behavior is that the master password is always cleared when the DB is successfully unlocked, but kept for you to fix if the password was entered incorrectly.  The issue I'm raising is that rotating the screen during a successful unlock results in the password not being cleared.  I am of the opinion that the password state after locking the DB should not be influenced by screen rotations that occurred while unlocking the DB.

This setting ensures that the master password is always cleared when you navigate away from the app.  So yes, enabling it will "fix up" the issue and prevent the damage of exposing the master password.  However, this setting is also affected by screen rotations.  When enabled, if you incorrectly type your password, and rotate the screen during unlock, the password is cleared, preventing you from fixing it and trying again.  Also when enabled, simply rotating the screen during password entry causes the password to be cleared.  These behaviors are unexpected, since I never left the password screen.

So, enabling this setting will address my security concern as a workaround, but it still seems to me that there is an underlying problem with how the app is handling screen rotations.

**Comment by 4-FLOSS-Free-Libre-Open-Source-Software** [comment]
> this can expose the master password by clicking the "eye".
oh, this is critical issue

**Comment by PhilippC** [comment]
can anybody reproduce this? I can't? (Which doesn't mean I don't believe the OP, I'd just like to find out when this occurs and when not)

**Comment by roberga** [comment]
I cannot reproduce the exact same issue as the OP but I can reproduce the related one mentioned in the bug report which I think can explain also the main issue in the bug report.

"Locking the DB via device screen lock is also affected. Password stays entered when app is sent to the background, other apps used, and later brought to the foreground."

If you switch to another app using the "Recent apps" button in Android or lock the screen, the setting "Clear entered master password" doesn't work.

The setting "Clear entered master password" is described as "Clear the entered master password when leaving the screen without unlocking the database". The entered master password is **not** cleared when switching to another app with the "Recent apps" button. When you later switch back to Keepass2Android, the previously entered password is still there. 

Switching to another app with "Recent Apps" button or locking the screen should clear the password in the same way as leaving the screen in other ways, such as with the back button. It is a security risk if you have typed the correct password in full or partially and it is not cleared if you switch to another app, eg. if you get a notification. Keepass2Android should clear any entered password when losing focus as foreground app.

**Comment by roberga** [comment]
BTW, what does the setting "Clear entered master password" do now? It seems that the master password is always cleared when leaving the app with the back button regardless of the value of this setting.
